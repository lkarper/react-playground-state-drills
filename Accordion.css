ul {
    list-style: none;
}

button {
    border: none;
    background-color: white;
    font-size: 20px;
}

button:hover, button:focus {
    text-decoration: underline;
}

button:focus {
    color: red;
}

p {
    font-weight: bold;
}