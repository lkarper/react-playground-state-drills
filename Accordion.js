import React from 'react';
import './Accordion.css';

class Accordion extends React.Component {
    static defaultProps = {
        sections: []
    }

    state = {
        index: null
    }

    handleClick = (i) => {
        if (this.state.index === i) {
            this.setState({
                index: null
            })
        } else {
            this.setState({
                index: i
            })
        }
    }

    render() {
        const accordionArray = this.props.sections.map((section, i) => {
            return (
                <li key={i}>
                    <button 
                        className={`button${i}`}
                        onClick={()=> this.handleClick(i)}
                    >
                        {section.title}
                    </button>
                    {this.state.index === i ? <p>{section.content}</p> : ''}
                </li>
            )
        });
        if (accordionArray.length === 0) {
            accordionArray.push(<li key={0}></li>)
        }
        return (
            <ul>
                {accordionArray}
            </ul>
        )
    }

}

export default Accordion;