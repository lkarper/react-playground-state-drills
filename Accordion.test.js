import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Accordion from './Accordion';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

const sections = [
    {
      title: 'Section 1',
      content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
    },
    {
      title: 'Section 2',
      content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
    },
    {
      title: 'Section 3',
      content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
    },
]

describe('Accordion component', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Accordion />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders an empty <li /> when the sections prop is not supplied', () => {
        const tree = renderer
            .create(<Accordion />)
            .toJSON();
            expect(tree).toMatchSnapshot();
    });

    it('renders no sections as active by default', () => {
        const tree = renderer
            .create(<Accordion sections={sections} />)
            .toJSON();
            expect(tree).toMatchSnapshot();
    });

    it('opens a clicked section', () => {
        const wrapper = shallow(<Accordion sections={sections} />);
        wrapper.find('.button0').simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('only opens one section at a time', () => {
        const wrapper = shallow(<Accordion sections={sections} />);
        wrapper.find('.button0').simulate('click');
        wrapper.find('.button2').simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();
    })

    it('closes a section that is clicked twice', () => {
        const wrapper = shallow(<Accordion sections={sections} />)
        wrapper.find('.button0').simulate('click');
        wrapper.find('.button0').simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();
    });
});