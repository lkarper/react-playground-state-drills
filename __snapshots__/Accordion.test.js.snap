// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component closes a section that is clicked twice 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      className="button0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      className="button1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      className="button2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion component only opens one section at a time 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      className="button0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      className="button1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      className="button2"
      onClick={[Function]}
    >
      Section 3
    </button>
    <p>
      Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
    </p>
  </li>
</ul>
`;

exports[`Accordion component opens a clicked section 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      className="button0"
      onClick={[Function]}
    >
      Section 1
    </button>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      className="button1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      className="button2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion component renders an empty <li /> when the sections prop is not supplied 1`] = `
<ul>
  <li />
</ul>
`;

exports[`Accordion component renders no sections as active by default 1`] = `
<ul>
  <li>
    <button
      className="button0"
      onClick={[Function]}
    >
      Section 1
    </button>
    
  </li>
  <li>
    <button
      className="button1"
      onClick={[Function]}
    >
      Section 2
    </button>
    
  </li>
  <li>
    <button
      className="button2"
      onClick={[Function]}
    >
      Section 3
    </button>
    
  </li>
</ul>
`;
